samples/javascript/index.js

```diff
@@ -0,0 +1,253 @@
+    createBuildEventHandlers(build) {
+        return this.eventHandlerClasses.reduce((handlers, handlerClass) => {
+            const addHandler = (type, handler) => handlers[type] ? handlers[type].push(handler) : handlers[type] = [handler];
+
+            const handler = new handlerClass.prototype.constructor(build);
+
+            this.getHandledEventTypesForHandlerClass(handlerClass).forEach(eventType => addHandler(eventType, handler));
+
+            if (Object.getOwnPropertyNames(handlerClass.prototype).includes('complete')) {
+                addHandler('complete', handler);
+            }
+
+            return handlers;
+        }, {});
+    }
+
+    processBuild(build) {
+        this.numBuildsInProcess++;
+        const buildEventHandlers = this.createBuildEventHandlers(build);
+        const buildEventStreamUrl = this.createBuildEventStreamUrl(build.buildId);
+
+        createServerSideEventStream(buildEventStreamUrl, {
+            oncomplete:  () => {
+                this.finishedProcessingBuild();
+
+                // Call the 'complete()' method on any handler that has it.
+                if (buildEventHandlers.complete) {
+                    buildEventHandlers.complete.forEach(handler => handler.complete());
+                }
+            },
+            eventListeners: [
+                {
+                    eventName: 'BuildEvent',
+                    eventHandler:event => {
+                        const buildEventPayload = JSON.parse(event.data);
+                        const { eventType } = buildEventPayload.type;
+
+                        if (this.allHandledEventTypes.has(eventType)) {
+                            buildEventHandlers[eventType].forEach(handler => handler[`on${eventType}`](buildEventPayload));
+                        }
+                    }
+                }
+            ],
+            retry: {
+                interval: 2000,
+                maxRetries: 100
+            }
+        });
+    }
+
+    finishedProcessingBuild() {
+        this.numBuildsInProcess--;
+        setTimeout(() => this.processPendingBuilds(), 0); // process the next set of pending builds, if any
+    }
+}
+
+// Code below is a wrapper of EventSourcePolyfill to provide an oncomplete callback,  retry interval and max retries configuration
+function createServerSideEventStream(url, configuration) {
+    const STATUS_COMPLETE = 204;
+
+    let stream;
+    let retries;
+
+    const _onopen = configuration.onopen || noop;
+    const _onerror = configuration.onerror || noop;
+    const _oncomplete = configuration.oncomplete || noop;
+    const _configurationRetry = configuration.retry || { }
+    const _maxRetries = _configurationRetry.maxRetries || 3;
+    const _retryInterval = _configurationRetry.interval || 1000;
+
+    stream = createStream()
+
+    function createStream() {
+        stream = new EventSourcePolyfill(url, { headers: {'Authorization': `Basic ${BASIC_AUTH_TOKEN}`}});
```

Comment:
Can we pass the `BASIC_AUTH_TOKEN` as a parameter instead of using it as a global constant?